hal/halx86/acpi/madt.c

```diff
@@ -12,77 +13,113 @@
-    HalpApicInfoTable.ApicMode = 0x10;
-    MadtTable = HalAcpiGetTable(LoaderBlock, 'CIPA');
 
-    AcpiHeader = (ACPI_SUBTABLE_HEADER*)MadtTable;
-    AcpiHeader->Length = sizeof(ACPI_TABLE_MADT);
-    TableEnd = (ULONG_PTR)MadtTable + MadtTable->Header.Length;
+    MadtTable = HalAcpiGetTable(LoaderBlock, APIC_SIGNATURE);
+    if (!MadtTable)
+    {
+        DPRINT1("HalAcpiGetTable(APIC) failed\n");
+        return;
+    }
 
-    HalpApicInfoTable.ProcessorCount = 0;
-    HalpProcessorIdentity = HalpStaticProcessorIdentity;
+    /* TODO: We only support legacy APIC for now */
+    HalpApicInfoTable.ApicMode = APIC_MODE_LEGACY;
 
-    AcpiHeader = (ACPI_SUBTABLE_HEADER*)MadtTable;
-    AcpiHeader->Length = sizeof(ACPI_TABLE_MADT);
+    AcpiHeader = (ACPI_SUBTABLE_HEADER *)((ULONG_PTR)MadtTable + sizeof(*MadtTable));
     TableEnd = (ULONG_PTR)MadtTable + MadtTable->Header.Length;
+    DPRINT("MadtTable %p, subtables %p - %08Ix\n", MadtTable, AcpiHeader, TableEnd);
 
-    while ((ULONG_PTR)AcpiHeader <= TableEnd)
+    while ((ULONG_PTR)AcpiHeader < TableEnd)
     {
-        LocalApic = (ACPI_MADT_LOCAL_APIC*)AcpiHeader;
-
-        if (LocalApic->Header.Type == ACPI_MADT_TYPE_LOCAL_APIC &&
-            LocalApic->Header.Length == sizeof(ACPI_MADT_LOCAL_APIC))
+        switch (AcpiHeader->Type)
         {
-            ValidProcessorCount = HalpApicInfoTable.ProcessorCount;
-
-            HalpProcessorIdentity[ValidProcessorCount].LapicId = LocalApic->Id;
-            HalpProcessorIdentity[ValidProcessorCount].ProcessorId = LocalApic->ProcessorId;
-
-            HalpApicInfoTable.ProcessorCount++;
-
-            AcpiHeader = (ACPI_SUBTABLE_HEADER*)((ULONG_PTR)AcpiHeader + AcpiHeader->Length);
-        }
-        else
-        {
-            /* End the parsing early if we don't use the currently selected table */
-            AcpiHeader = (ACPI_SUBTABLE_HEADER*)((ULONG_PTR)AcpiHeader + 1);
+            case ACPI_MADT_TYPE_LOCAL_APIC:
+            {
+                ACPI_MADT_LOCAL_APIC *LA = (ACPI_MADT_LOCAL_APIC *)AcpiHeader;
+
+                DPRINT(" Local Apic, Processor %lu: ProcessorId %u, Id %u, LapicFlags %08X\n",
+                       HalpApicInfoTable.ProcessorCount,
+                       LA->ProcessorId, LA->Id, LA->LapicFlags);
+
+                HalpProcessorIdentity[HalpApicInfoTable.ProcessorCount].ProcessorId = LA->ProcessorId;
+                HalpProcessorIdentity[HalpApicInfoTable.ProcessorCount].LapicId = LA->Id;
+#if 0 // TODO: Is that correct?
+                if (LA->LapicFlags & 0x00000001) // Enabled
+                    HalpProcessorIdentity[HalpApicInfoTable.ProcessorCount].ProcessorStarted = TRUE;
+#endif
+
+                HalpApicInfoTable.ProcessorCount++;
+
+                break;
+            }
+            case ACPI_MADT_TYPE_IO_APIC:
+            {
+#if DBG && !defined(NDEBUG)
+                ACPI_MADT_IO_APIC *IA = (ACPI_MADT_IO_APIC *)AcpiHeader;
```

Comment:
These shortened names are a little hard to read. 
Why not something like IoApic?